tests/bmtestspolar/symfixedrankpolarfactory.m

```diff
@@ -0,0 +1,264 @@
+				+ trinner(Binvetazeta(:, 1:k), Binvetazeta(:, 1+k : 2*k));
+    end
+    
+    M.norm = @(X, eta) sqrt(M.inner(X, eta, eta));
+    
+    M.dist = @(x, y) error('symfixedrankpolarfactory.dist not implemented yet.');
+    
+    M.typicaldist = @() 10*k;
+    
+    skew = @(X) .5*(X-X');
+    symm = @(X) .5*(X+X');
+    stiefel_proj = @(U, H) H - U*symm(U'*H);
+
+	% It is sometimes useful to switch between representation of matrices
+	% as pairs or as full matrices of size m x m. The function to
+	% convert a matrix to a pair, matrix2pair, allows to specify the
+	% rank of the representation. By default, it is equal to k. Omit the
+	% second input (or set to inf) to get a full EVD pair (in economy
+	% format). If so, the resulting pair does not represent a point on
+	% the manifold.
+	M.matrix2pair = @matrix2pair;
+	function X_pair = matrix2pair(X_matrix, r)
+	    if ~exist('r', 'var') || isempty(r) || r <= 0
+	        r = k;
+	    end
+	    if r < min(m, n)
+	        [U, B] = eigs(X_matrix, r);
+	    else
+	        [U, B] = eig(X_matrix);
+	    end
+	    X_pair.U = U;
+	    X_pair.B = B;
+	end
+	M.pair2matrix = @pair2matrix;
+	function X_matrix = pair2matrix(X_pair)
+	    U = X_pair.U;
+	    B = X_pair.B;
+	    X_matrix = U*B*U';
+	end
+    
+    M.egrad2rgrad = @egrad2rgrad;
+    function rgrad = egrad2rgrad(X, egrad)
+        rgrad.U = stiefel_proj(X.U, egrad.U);
+        rgrad.B = X.B*symm(egrad.B)*X.B;
+    end
+    
+    
+    M.ehess2rhess = @ehess2rhess;
+    function Hess = ehess2rhess(X, egrad, ehess, eta)
+        % Mathematical development of the below computation is
+        % as follows.
+        % We first compute the Riemannian gradient for the factor B.
+        % rgrad.B = X.B*symm(egrad.B)*X.B;
+        %
+        % Directional derivatives of the Riemannian gradient.
+        % Hess.U = ehess.U - eta.U*symm(X.U'*egrad.U);
+        % Hess.U = stiefel_proj(X.U, Hess.U);
+        %
+        % Hess.B = X.B*symm(ehess.B)*X.B +  2*symm(eta.B*symm(egrad.B)*X.B);
+        %
+        % Finally, a correction factor for the non-constant metric on the factor B.
+        % Hess.B = Hess.B - symm(eta.B*(X.B\rgrad.B));
+        Hess.U = ehess.U - eta.U*symm(X.U'*egrad.U);
+        Hess.U = stiefel_proj(X.U, Hess.U);
+
+        Hess.B = X.B*symm(ehess.B)*X.B + symm(eta.B*symm(egrad.B)*X.B);
+
+		
+        % Projection onto the horizontal space.
+        Hess = M.proj(X, Hess);
+    end
+    
+    
+    M.proj = @projection;
+    function etaproj = projection(X, eta)
+        % First, projection onto the tangent space of the total space.
+        eta.U = stiefel_proj(X.U, eta.U);
+        eta.B = symm(eta.B);
+        
+        % Then, projection onto the horizontal space.
+        SS = X.B*X.B';
+        AS = X.B*(skew(X.U'*eta.U) - 2*skew(X.B\eta.B))*X.B;
+        
+        % Compute skew-symmetric Omega.
+        % To solve the system SS*Omega + Omega*SS - B*Omega*B = AS
+        Omega = mylinearsystem(SS, AS);
```

Comment:
```suggestion
        % TODO: NB, May 3, 2019:
        % 
		% If we apply inv(B) ... inv(B) to this equation, then inv(B)ASinv(B)
        % simplifies to just its middle part (skew symmetric) (call it Z here),
        % and we have
        % 
        % B Omega inv(B) + inv(B) * Omega B - Omega = Z
        % 
        % Should we be solving this equation in this form rather than the
        % current one? We can do the same trick of computing [V, D] = eig(B),
        % and substituting VDV' in the equation, to get a simple form, where
        % 
        % S = V' Omega V
        % 
        % and S(i, j) = Z(i, j) / (D(i, i)/D(j, j) + D(j, j)/D(i, i) - 1)
        % 
        % which can surely be computed without for-loops.
        % 
        % In any case, I don't know if it would be much better, but it does
        % seem to save some computations, in particular the squaring of B
        % (which may be bad for conditioning) and hitting AS with B on both sides.
        Omega = mylinearsystem(SS, AS);
```

Adding the comment as a TODO in the code so we don't lose track of it. The squaring worries me, because of conditioning.